cpp/src/lib/rora_proxy_client.cc

```diff
@@ -154,53 +166,154 @@ void RoraProxy_client::_maybe_update_osd_infos(
   }
 }
 
-int RoraProxy_client::_short_path_one(const std::string & /*namespace_*/,
-                                      const ObjectSlices &object_slices,
-                                      std::shared_ptr<Manifest> mfp) {
+Location get_location(ManifestWithNamespaceId &mf, uint64_t pos, uint32_t len) {
+  int chunk_index = -1;
+  uint64_t total = 0;
+
+  {
+    auto it = mf.chunk_sizes.begin();
+    while (total <= pos) {
+      chunk_index++;
+      auto chunk_size = *it;
+      total += chunk_size;
+      it++;
+    }
+  }
 
-  // one object, maybe multiple slices and or fragments involved
-  std::map<osd_t, std::vector<asd_slice>> per_osd;
-  const Manifest &manifest = *mfp;
-  for (auto &sd : object_slices.slices) {
-    uint32_t bytes_to_read = sd.size;
-    uint32_t offset = sd.offset;
-    byte *buf = sd.buf;
-
-    while (bytes_to_read > 0) {
-      auto maybe_coords = manifest.to_chunk_fragment(offset);
-      if (maybe_coords == boost::none) {
-        return -1;
-      }
+  auto &chunk_fragment_locations = mf.fragment_locations[chunk_index];
 
-      auto &coords = *maybe_coords;
+  uint32_t chunk_size = mf.chunk_sizes[chunk_index];
+  total -= chunk_size;
+  uint32_t fragment_length = chunk_size / mf.encoding_scheme.k;
+  uint32_t pos_in_chunk = pos - total;
 
-      osd_t osd = coords._osd;
-      auto it = per_osd.find(osd);
-      if (it == per_osd.end()) {
-        std::vector<asd_slice> slices;
-        per_osd.insert(make_pair(osd, slices));
-        it = per_osd.find(osd);
-      }
+  uint32_t fragment_index = pos_in_chunk / fragment_length;
+  auto p = chunk_fragment_locations[fragment_index];
 
-      uint32_t bytes_in_slice;
-      if (coords.pos_in_fragment + bytes_to_read <= coords.fragment_length) {
-        bytes_in_slice = bytes_to_read;
-      } else {
-        bytes_in_slice = coords.fragment_length - coords.pos_in_fragment;
+  total += fragment_length * fragment_index;
+  uint32_t pos_in_fragment = pos - total;
+
+  Location l;
+  l.namespace_id = mf.namespace_id;
+  l.object_id = mf.object_id;
+  l.chunk_id = chunk_index;
+  l.fragment_id = fragment_index;
+  l.fragment_location = p;
+  l.offset = pos_in_chunk;
+  l.length = std::min(len, fragment_length - pos_in_fragment);
+  return l;
+}
+
+void _resolve_slice_one_level(std::vector<std::pair<byte *, Location>> &results,
+                              ManifestWithNamespaceId &manifest,
+                              uint64_t offset, uint32_t length, byte *target) {
+  while (length > 0) {
+    Location l = get_location(manifest, offset, length);
+    length -= l.length;
+    offset += l.length;
+    results.push_back(std::move(std::pair<byte *, Location>(target, l)));
```

Comment:
As an exercise in code golf: :)

AFAICS `Location` does not have a move ctor / assigment operator, so there will be
* a copy into the temporary pair
* a copy for the intended move into the vector (btw, the constructed `pair` is already a "temporary" so there's no need for the `std::move`).
.
The latter could for example be avoided by using the `vector`'s `emplace` to construct the pair right into the vector - something along those lines:
```
results.emplace_back(target, get_location(...));
length -= results.back().length;
offset += results.back().length;
```
Taking a step back, the `results` vector is possibly resized, which will lead to more copies. This could be avoided by doing a `reserve` to the right size beforehand.